plugins/bcm57xx/fu-bcm57xx-ethtool-device.c

```diff
@@ -0,0 +1,534 @@
+ * Copyright (C) 2018-2020 Evan Lojewski
+ * Copyright (C) 2020 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: GPL-2+
+ */
+
+#include "config.h"
+
+#ifdef HAVE_ETHTOOL_H
+#include <linux/ethtool.h>
+#include <linux/sockios.h>
+#include <net/if.h>
+#endif
+#ifdef HAVE_IOCTL_H
+#include <sys/ioctl.h>
+#endif
+#ifdef HAVE_SOCKET_H
+#include <sys/socket.h>
+#endif
+
+#include "fu-common.h"
+
+#include "fu-bcm57xx-common.h"
+#include "fu-bcm57xx-ethtool-device.h"
+#include "fu-bcm57xx-mmap-device.h"
+#include "fu-bcm57xx-firmware.h"
+
+struct _FuBcm57xxEthtoolDevice {
+	FuUdevDevice		 parent_instance;
+	gchar			*ethtool_iface;
+	int			 ethtool_fd;
+};
+
+G_DEFINE_TYPE (FuBcm57xxEthtoolDevice, fu_bcm57xx_ethtool_device, FU_TYPE_UDEV_DEVICE)
+
+static void
+fu_bcm57xx_ethtool_device_to_string (FuUdevDevice *device, guint idt, GString *str)
+{
+	FuBcm57xxEthtoolDevice *self = FU_BCM57XX_ETHTOOL_DEVICE (device);
+	fu_common_string_append_kv (str, idt, "EthtoolIface", self->ethtool_iface);
+}
+
+static gboolean
+fu_bcm57xx_ethtool_device_probe (FuUdevDevice *device, GError **error)
+{
+	FuBcm57xxEthtoolDevice *self = FU_BCM57XX_ETHTOOL_DEVICE (device);
+	g_autofree gchar *fn = NULL;
+	g_autoptr(GPtrArray) ifaces = NULL;
+
+	/* only enumerate number 1 */
+	if (fu_udev_device_get_number (device) != 1) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "only device 1 supported on multi-device card");
+		return FALSE;
+	}
+
+	/* only if has an interface */
+	fn = g_build_filename (fu_udev_device_get_sysfs_path (device), "net", NULL);
+	ifaces = fu_common_filename_glob (fn, "en*", NULL);
+	if (ifaces == NULL || ifaces->len == 0) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "no interface detected");
+		return FALSE;
+	}
+	self->ethtool_iface = g_path_get_basename (g_ptr_array_index (ifaces, 0));
+
+	/* success */
+	return fu_udev_device_set_physical_id (device, "pci", error);
+}
+
+static gboolean
+fu_bcm57xx_ethtool_device_nvram_write (FuBcm57xxEthtoolDevice *self,
+				       guint32 address,
+				       const guint32 *buf,
+				       guint32 bufsz_wrds,
+				       GError **error)
+{
+#ifdef HAVE_ETHTOOL_H
+	gsize eepromsz;
+	gint rc = -1;
+	struct ifreq ifr = { 0 };
+	g_autofree struct ethtool_eeprom *eeprom = NULL;
+
+	/* sanity check */
+	if (address + bufsz_wrds * sizeof(guint32) > fu_device_get_firmware_size_max (FU_DEVICE (self))) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "tried to read outside of EEPROM size [0x%x]",
+			     (guint) fu_device_get_firmware_size_max (FU_DEVICE (self)));
+		return FALSE;
+	}
+
+	/* write EEPROM (NVRAM) data */
+	eepromsz = sizeof(struct ethtool_eeprom) + bufsz_wrds * sizeof(guint32);
+	eeprom = (struct ethtool_eeprom *) g_malloc0 (eepromsz);
+	eeprom->cmd = ETHTOOL_SEEPROM;
+	eeprom->len = bufsz_wrds * sizeof(guint32);
+	eeprom->offset = address;
```

Comment:
This returns -1 if the following isn't added:
    eeprom->magic = BCM_NVRAM_MAGIC;

Reply:
Indeed, thanks!